### PR TITLE
Perform obligation deduplication to avoid buggy `ExistentialMismatch`

### LINKED ISSUE #59326: Type mismatching cased by duplicate associated type resolution
Apologize that I could not give any code example right now. I'm working on creating one but it seems tricky. 

The compiler is complaining about `[E0308]: mismatched types` where it said it's expecting for a trait `SomeTrait<Apple=SomeApple, Apple=SomeApple, Banana=SomeBanana, Banana=SomeBanana>` but got `SomeTrait<Apple=SomeAPple, Banana=SomeBanana>`. Only some of the associated types in that trait are repeated once and there are few others are not repeated.

The example I gave in #59324 is an intermediate result of my attempt to create a minimal reproducible code to this issue. The type mismatch happened at the `with_factory` method when resolving concrete type for `ThriftService`.

The error happened when I'm doing the update from 1.32.0 to 1.33.0 so it's clearly a regression. 

Also, please let me know if you have any suggestion on bypassing this problem or tips on re-creating the problematic code.

Thanks.

### PR DESCRIPTION
**Pull request by estebank**
Address #59326.

### CODE REVIEW

File: src/librustc_middle/ty/relate.rs [hunk 1]
```diff
@@ -617,12 +617,22 @@ impl<'tcx> Relate<'tcx> for &'tcx ty::List<ty::ExistentialPredicate<'tcx>> {
...
-        if a.len() != b.len() {
+        let tcx = relation.tcx();
+
+        // FIXME: this is wasteful, but want to do a perf run to see how slow it is.
+        // We need to perform this deduplication as we sometimes generate duplicate projections
+        // in `a`.
+        let mut a_v: Vec<_> = a.into_iter().collect();
+        let mut b_v: Vec<_> = b.into_iter().collect();
+        a_v.sort_by(|a, b| a.stable_cmp(tcx, b));
+        a_v.dedup();
```
Comment (hellow554) [inline src/librustc_middle/ty/relate.rs:628]: Is it correct, that the first one is a `sort_by`, but the next one a plain `dedup`, instead of `dedup_by`?

Comment (estebank) [reply src/librustc_middle/ty/relate.rs:628]: You're right, that is probably a potential bug. On the repro case that I posted it didn't cause problems, but I'll change it after the perf run has completed (don't want to break anything by force-pushing to this branch before then).

Comment (estebank) [reply src/librustc_middle/ty/relate.rs:628]: I'm not sure if `dedup` will cause trouble in practice and don't know whether `dedup_by_key` might introduce a slight slowdown. Either way, this PR seems to solve bad rejection part of the ticket.

**Comment by spastorino** [comment]
Adding `T-compiler` so our tools are able to fetch it and include it on the weekly agenda.

**Comment by nikomatsakis** [comment]
@bors r+

As I said on Zulip, this doesn't seem like the "right fix" to me, but it is a fix, and I don't see it causing any fresh problems, and there's an imminent regression, so r+.

File: src/librustc_middle/ty/relate.rs [hunk 2]
```diff
@@ -617,12 +617,22 @@ impl<'tcx> Relate<'tcx> for &'tcx ty::List<ty::ExistentialPredicate<'tcx>> {
         a: &Self,
         b: &Self,
     ) -> RelateResult<'tcx, Self> {
-        if a.len() != b.len() {
+        let tcx = relation.tcx();
+
+        // FIXME: this is wasteful, but want to do a perf run to see how slow it is.
```
Comment (nikomatsakis) [inline src/librustc_middle/ty/relate.rs:622]: I wonder if we should open up a FIXME issue for this..? To try and prevent the duplicates in the first place and figure out where they come from?

**Comment by pnkfelix** [comment]
discussed in T-compiler [meeting](https://zulip-archive.rust-lang.org/238009tcompilermeetings/59499weeklymeeting2020062554818.html#201976290). Approved for beta-backport.

**Comment by pnkfelix** [comment]
stable nom was also [discussed](https://zulip-archive.rust-lang.org/238009tcompilermeetings/59499weeklymeeting2020062554818.html#201978099). Stable backport approved.

**Pull request merged**